### PR TITLE
Azure Firewall Cmdlets

### PR DESCRIPTION
**Pull request by joannakleong**
## Description

Adding new cmdlets for Azure firewall
Design review found here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/87

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by praries880** [comment]
@joannakleong The new tests are failing in the Jenkins build :
```
22:35:55 (InvokeXUnit target) -> 
22:35:55   c:\workspace\powershell\src\ResourceManager\Network\Commands.Network.Test\NetworkResourcesController.cs(86): error : Commands.Network.Test.ScenarioTests.AzureFirewallTests.TestAzureFirewallCRUD [FAIL] [c:\workspace\powershell\build.proj]
22:35:55   c:\workspace\powershell\src\ResourceManager\Network\Commands.Network.Test\NetworkResourcesController.cs(86): error : Commands.Network.Test.ScenarioTests.AzureFirewallTests.TestAzureFirewallAllocateAndDeallocate [FAIL] [c:\workspace\powershell\build.proj]
22:35:55 
```

**Comment by joannakleong** [comment]
@praries880 is it possible to review the code while in the meantime I see why the test is failing? It will take a few hours and I don't want this to block code review

**Comment by joannakleong** [comment]
@praries880 The tests pass for me locally, I believe they are failing in Jenkins build because it is not picking up the correct "Microsoft.Azure.Management.Network" dll based off of the NuGet I placed in /LocalFeed folder. Do I need to also commit this dll?

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            Mandatory = true,
+            HelpMessage = "The protocols of the rule")]
+        [ValidateNotNullOrEmpty]
+        public List<string> Protocol { get; set; }
+        
+        public override void Execute()
+        {
+            base.Execute();
+
+            if (this.Protocol == null || this.Protocol.Count == 0)
```
[the remark this answers is not in the document]
Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/ApplicationRule/NewAzureFirewallApplicationRuleCommand.cs]: I see, thanks, good to know!

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                Description = this.Description,
+                Protocols = this.Protocol,
+                SourceAddresses = this.SourceAddress,
+                DestinationAddresses = this.DestinationAddress,
+                DestinationPorts = this.DestinationPort
+            };
+            WriteObject(networkRule);
+        }
+
+        private void ValidateProtocols(List<string> protocols)
```
[the remark this answers is not in the document]
Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/NetworkRule/NewAzureFirewallNetworkRuleCommand.cs]: cool, thanks!

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            HelpMessage = "Do not ask for confirmation if you want to overrite a resource")]
+        public SwitchParameter Force { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+        public SwitchParameter AsJob { get; set; }
+
+        public override void Execute()
+        {
+            var isVnetPresent = !string.IsNullOrEmpty(VirtualNetworkName);
+            var isPublicIpPresent = !string.IsNullOrEmpty(PublicIpName);
```
[the remark this answers is not in the document]
Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs]: interesting, good to know

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            {
+                var vnet = this.VirtualNetworkClient.Get(this.ResourceGroupName, VirtualNetworkName);
+                this.virtualNetwork = NetworkResourceManagerProfile.Mapper.Map<PSVirtualNetwork>(vnet);
+                
+                var publicIp = this.PublicIPAddressesClient.Get(this.ResourceGroupName, PublicIpName);
+                this.publicIpAddress = NetworkResourceManagerProfile.Mapper.Map<PSPublicIpAddress>(publicIp);
+            }
+            
+            base.Execute();
+            WriteWarning("The output object type of this cmdlet will be modified in a future release.");
```
[the remark this answers is not in the document]
Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs]: Actually we don't need this warning, removing it :)

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+        [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+        public SwitchParameter AsJob { get; set; }
+
+        public override void Execute()
+        {
+            var isVnetPresent = !string.IsNullOrEmpty(VirtualNetworkName);
+            var isPublicIpPresent = !string.IsNullOrEmpty(PublicIpName);
+
+            if ((isVnetPresent && !isPublicIpPresent) || (!isVnetPresent && isPublicIpPresent))
```
[the remark this answers is not in the document]
Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs]: I've created a parameter set for VirtualNetworkName and PublicIpName, but I haven't set the default parameter set yet. How does the default parameter set work? Do I just add the label for DefaultParameterSetName for the cmdlet and also all the remaining parameters?

**Comment by praries880** [comment]
@joannakleong Netcode build is broken with : https://travis-ci.org/Azure/azure-powershell/builds/397850479#L1354, which means the net core build is not able to find th library.

Jenkins build is broken because it did not find a change to the changelog file. : 
```
08:34:09 c:\workspace\powershell\build.proj(613,5): error : Modified files were found with no update to their change log. Please add a snippet to the affected modules' change log.
```

File: src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        [Parameter(Mandatory = false, HelpMessage = "Run cmdlet in the background")]
+        public SwitchParameter AsJob { get; set; }
+
+        public override void Execute()
+        {
+            var isVnetPresent = VirtualNetworkName==null;
+            
+            // Get the virtual network, get the public IP address
+            if (isVnetPresent)
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs]: look at https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Billing/Commands.Billing/Cmdlets/BillingPeriods/GetAzureRmBillingPeriod.cs#L39for how to see which param set is used and using the appropriate parmas from the param set.

Comment (joannakleong) [reply src/ResourceManager/Network/Commands.Network/AzureFirewall/NewAzureFirewallCommand.cs]: thanks :) I believe I've fixed it now

**Comment by joannakleong** [comment]
@praries880 the Travis CI build is failing with error: "The term 'New-AzureRmFirewall' is not recognized as the name of a cmdlet, function, script file, or operable program." - I'm not sure how to fix this as the tests are running fine for me locally. Any ideas?

**Comment by joannakleong** [comment]
@praries880 finally all checks have passed!

**Comment by praries880** [comment]
on demand : https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/970/

**Pull request merged**